Sources/Plasma/FeatureLib/pfGLPipeline/plGLEnumerate.cpp

```diff
@@ -0,0 +1,282 @@
+        EGL_NONE
+    };
+    EGLContext context = eglCreateContext(display, configs[0], EGL_NO_CONTEXT, ctx_attrs);
+    if (context == EGL_NO_CONTEXT)
+        return;
+
+    EGLSurface surface = eglCreatePbufferSurface(display, configs[0], nullptr);
+    if (surface == EGL_NO_SURFACE)
+        return;
+
+    if (!eglMakeCurrent(display, surface, surface, context))
+        return;
+
+    hsG3DDeviceRecord devRec;
+    devRec.SetG3DDeviceType(hsG3DDeviceSelector::kDevTypeOpenGL);
+    devRec.SetDriverName("EGL");
+
+    if (fillDeviceRecord(devRec))
+        records.emplace_back(devRec);
+
+    eglMakeCurrent(display, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+    eglDestroySurface(display, surface);
+    eglDestroyContext(display, context);
+    eglTerminate(display);
+}
+#endif // USE_EGL
+
+
+#ifdef HS_BUILD_FOR_WIN32
+#include "hsWindows.h"
+#include <tchar.h>
+#include <epoxy/wgl.h>
+
+void cleanupWGL(LPCTSTR cls, HWND wnd = 0, HDC dc = 0, HGLRC ctx = 0)
+{
+    if (ctx) {
+        wglMakeCurrent(nullptr, nullptr);
+        wglDeleteContext(ctx);
+    }
+
+    if (dc) {
+        ReleaseDC(wnd, dc);
+    }
+
+    if (wnd) {
+        DestroyWindow(wnd);
+    }
+
+    UnregisterClass(cls, GetModuleHandle(nullptr));
+}
+
+void plWGLEnumerate(std::vector<hsG3DDeviceRecord>& records)
+{
+    WNDCLASS tempClass;
+    memset(&tempClass, 0, sizeof(tempClass));
+    tempClass.lpfnWndProc = DefWindowProc;
+    tempClass.hInstance = GetModuleHandle(nullptr);
+    tempClass.hbrBackground = (HBRUSH)GetStockObject(BLACK_BRUSH);
+    tempClass.lpszClassName = _T("GLTestClass");
+    tempClass.style = CS_OWNDC;
+
+    ATOM cls = RegisterClass(&tempClass);
+    if (!cls)
+        return;
+
+    LPCTSTR clsName = reinterpret_cast<LPCTSTR>(cls);
+
+    HWND wnd = CreateWindowEx(0, clsName, _T("OpenGL Test Window"),
+                WS_POPUP | WS_CLIPSIBLINGS | WS_CLIPCHILDREN | WS_VISIBLE,
+                CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT,
+                nullptr, nullptr, GetModuleHandle(nullptr), nullptr);
+    if (!wnd)
+        return cleanupWGL(clsName);
+
+    HDC dc = GetDC(wnd);
+    if (!dc)
+        return cleanupWGL(clsName, wnd);
+
+    PIXELFORMATDESCRIPTOR pfd = {};
+    pfd.nSize = sizeof(PIXELFORMATDESCRIPTOR);
+    pfd.nVersion = 1;
+    pfd.dwFlags = PFD_DRAW_TO_WINDOW | PFD_GENERIC_ACCELERATED | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER;
+    pfd.iPixelType = PFD_TYPE_RGBA;
+    pfd.cColorBits = 24;
+    pfd.cDepthBits = 24;
+    pfd.iLayerType = PFD_MAIN_PLANE;
+
+    int format = ChoosePixelFormat(dc, &pfd);
+    if (!format)
+        return cleanupWGL(clsName, wnd, dc);
+
+    if (!SetPixelFormat(dc, format, &pfd))
+        return cleanupWGL(clsName, wnd, dc);
+
+    HGLRC ctx = wglCreateContext(dc);
+    if (!ctx)
+        return cleanupWGL(clsName, wnd, dc);
+
+    if (!wglMakeCurrent(dc, ctx))
+        return cleanupWGL(clsName, wnd, dc, ctx);
```

Comment:
This repeated cleanup logic could probably be cleaned up with smart pointers or do... while(0)